cpp/src/io/parquet/page_string_decode.cu

```diff
@@ -0,0 +1,1125 @@
+    using cudf::detail::warp_size;
+    auto const use_char_ll =
+      s->page.num_valids > 0 && (s->page.str_bytes / s->page.num_valids) >= warp_size;
+
+    auto [ptr, len] = src_pos < target_pos && dst_pos >= 0
+                        ? gpuGetStringData(s, sb, src_pos + skipped_leaf_values)
+                        : cuda::std::pair<char const*, size_t>{nullptr, 0};
+
+    using block_scan = cub::BlockScan<size_type, decode_block_size>;
+    __shared__ typename block_scan::TempStorage scan_storage;
+    size_type offset;
+    block_scan(scan_storage).ExclusiveSum(len, offset);
+
+    offset += last_offset;
+
+    if (use_char_ll) {
+      __shared__ __align__(8) uint8_t const* pointers[decode_block_size];
+      __shared__ __align__(4) size_type offsets[decode_block_size];
+      __shared__ __align__(4) int dsts[decode_block_size];
+      __shared__ __align__(4) int lengths[decode_block_size];
+
+      offsets[t]  = offset;
+      pointers[t] = reinterpret_cast<uint8_t const*>(ptr);
+      dsts[t]     = dst_pos;
+      lengths[t]  = len;
+      __syncthreads();
+
+      using cudf::detail::warp_size;
+      constexpr int nwarp = decode_block_size / warp_size;
+      int const warpid    = t / warp_size;
+      int const lane_id   = t % warp_size;
+      for (int ss = warpid; ss < decode_block_size && ss + s->src_pos < target_pos; ss += nwarp) {
+        if (dsts[ss] >= 0) {
+          auto offptr =
+            reinterpret_cast<int32_t*>(nesting_info_base[leaf_level_index].data_out) + dsts[ss];
+          *offptr      = lengths[ss];
+          auto str_ptr = nesting_info_base[leaf_level_index].string_out + offsets[ss];
+          ll_strcpy(str_ptr, pointers[ss], lengths[ss], lane_id);
+        }
+      }
+    } else {
+      if (src_pos < target_pos && dst_pos >= 0) {
+        auto offptr =
+          reinterpret_cast<int32_t*>(nesting_info_base[leaf_level_index].data_out) + dst_pos;
+        *offptr      = len;
+        auto str_ptr = nesting_info_base[leaf_level_index].string_out + offset;
+        memcpy(str_ptr, ptr, len);
+      }
+    }
+    __syncthreads();
+
+    // last thread in block updates last_offset.
+    if (t == decode_block_size - 1) {
+      last_offset = offset + len;
+      *(volatile int32_t*)&s->src_pos += decode_block_size;
+    }
+    __syncthreads();
+  }
+
+  // now turn array of lengths into offsets
+  int value_count = nesting_info_base[leaf_level_index].value_count;
+
+  // if no repetition we haven't calculated start/end bounds and instead just skipped
+  // values until we reach first_row. account for that here.
+  if (!has_repetition) { value_count -= s->first_row; }
+
+  auto const offptr = reinterpret_cast<size_type*>(nesting_info_base[leaf_level_index].data_out);
+  block_excl_sum<decode_block_size>(offptr, value_count, s->page.str_offset);
+}
+
+__global__ void __launch_bounds__(preprocess_block_size)
+  gpuComputePageOffsets(device_span<ColumnChunkDesc const> chunks, device_span<size_t> col_sizes)
```

Comment:
This smells like something that could be done as a `thrust::exclusive_scan_by_key()`.   We did something just like this recently to eliminate a similar nested loop:

https://github.com/rapidsai/cudf/blob/0c68ca15011774541985f46fb566e07c634e1615/cpp/src/io/parquet/reader_impl_preprocess.cu#LL1754C2-L1754C2

I think if you structured your keys as:

`A A A A A B B B B B C C C C C ...`

where

```
|-------|  size of each set of keys = (# of chunks * max # of pages in any chunk)
A A A A A

and the # of sets would be the number of columns.
```

```
elements_per_key = chunks.size() * (max pages in any chunk)
num_keys = elements_per_key * num_columns
key_iter(int index) = {
   return index / elements_per_key  //  col_index
}
value_iter(int index) = {
   // the logic here may not be exactly right but you get the idea
   col_index = index / elements_per_key        // col_index

   intra_index = index % elements_per_key    // index within each chunk/page group
   chunk_index = intra_index / max pages     // chunk index
   chunk = chunks[chunk_index];
   page_index = intra_index % max pages     // page index
  
   if(!is_string_col(chunk)){ return 0; }
   if(page_index > chunk.max_num_pages) { return 0; }
   if(chunk.pages[page_index].flags & DICTIONARY) { return 0; }
   return chunk.page_info[page_index].str_bytes;
}
thrust::scan_by_key(key_iter, key_iter + num_keys, value_iter, output)
```

After the scan you should have what you're looking for, and the whole thing becomes fully parallel - no nested looping.

Reply:
🤯 

Thanks!  I tried and failed to come up with a way to do this from the host side with thrust calls. :(  I'll see if I can get this to work. (Might be bugging you offline)

Reply:
@nvdbaranec please check what I've done. In particular the change [here](https://github.com/etseidl/cudf/blob/59bfd1be94bba846599c8989d0a91f338ee558b8/cpp/src/io/parquet/reader_impl_preprocess.cu#L1259). When I tried reusing `page_keys`, the ReorderedColumns reader test failed.  Switching from `src_col_schema` to `src_col_index` fixed that.